### ISSUE TITLE
EmberServer SecureSession Request Attribute Does Not Contain Expected Values

### ISSUE THREAD

**Issue by geoffreytchew** [http4s/http4s#5110]
When I migrated from Blaze Server to Ember Server (http4s version 0.23.1), I encountered an issue where I could no longer retrieve the client's certificates used for TLS communication from the SecureSession request attribute. An empty list is returned instead of the expected certificates. Upon further inspection it seems that all SecureSession field values are empty/default values.
Consider the following code:
```
    val secureSessionInfo = req.attributes.lookup(ServerRequestKeys.SecureSession).flatten
    secureSessionInfo match {
      case Some(secureSession) =>
        println(s"@!@ CERT LIST SIZE   = ${secureSession.X509Certificate.size}")
        println(s"@!@ SSL Session ID   = ${secureSession.sslSessionId}")
        println(s"@!@ SSL Cipher Suite = ${secureSession.cipherSuite}")
        println(s"@!@ SSL Key Size     = ${secureSession.keySize}")
      case None =>
        println("Secure session information could not be obtained from request")
    }
```
When using Ember Server, the following values are printed out:
```
server @!@ CERT LIST SIZE   = 0
server @!@ SSL Session ID   =
server @!@ SSL Cipher Suite = SSL_NULL_WITH_NULL_NULL
server @!@ SSL Key Size     = 0
```
When using Blaze Server, the following values are printed out:
```
server @!@ CERT LIST SIZE   = 3
server @!@ SSL Session ID   = 097023bbc36f7b19bd52536aeabcf880de7f19e3cb00e2b7a942b14e1f9a6c02
server @!@ SSL Cipher Suite = TLS_AES_256_GCM_SHA384
server @!@ SSL Key Size     = 0
```

**Comment by geoffreytchew** [comment]
I was able to reproduce this issue in unit tests that are based on `blaze-server/src/test/scala/org/http4s/blaze/server/BlazeServerMtlsSpec.scala`
I've attached these ember server unit tests in this comment.
To run the tests, do the following:
1. Unzip [EmberServerMtlsSuite.scala.zip](https://github.com/http4s/http4s/files/7078265/EmberServerMtlsSuite.scala.zip) and copy to `ember-server/src/test/scala/org/http4s/ember/server/EmberServerMtlsSuite.scala`
2. Copy `blaze-server/src/test/resources/keystore.jks` to `ember-server/src/test/resources/keystore.jks`
3. Run tests in ember-server
4. The new tests will fail because the SecureSession field values are unexpected (empty session id, 0 length key size)

Comparison between Blaze Server unit tests and Ember Server unit test values:
Ember Server tests:
!!! CERT LIST SIZE   = 0
!!! SSL Session ID   =
!!! SSL Cipher Suite = SSL_NULL_WITH_NULL_NULL
!!! SSL Key Size     = 0

Blaze Server tests:
!!! CERT LIST SIZE   = 2
!!! SSL Session ID   = 9642bac82fd51b63b28fdd243d67a8595ac0d0c2a3eeac08ebaa8894c7a7f845
!!! SSL Cipher Suite = TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384
!!! SSL Key Size     = 256

**Comment by mpilquist** [comment]
The problem is that ember-server is creating a `SecureSession` immediately after accepting a client connection. At that time, a snapshot of the socket's TLS state is captured and stored in the request vault. At this time, the TLS handshake hasn't yet begun, hence the state seen in the ember tests.

To fix, a `SecureSession` should be put in to the vault after reading the headers of *each* request -- note doing it on each request is important in case the handshake is renegotiated.

**Comment by geoffreytchew** [comment]
I'd like to submit a PR for the fix. I'm not able to assign this to myself though.

**Comment by ChristopherDavenport** [comment]
Fixed by #5138
